### DISCUSSION TITLE
Throwing custom exception

### DISCUSSION THREAD

**Discussion by siganberg** [ardalis/GuardClauses#163] [Ideas]
It would be nice to have an overloaded method to throw a custom exception instead of a default ArgumentException. Some of the code validation in our application requires us to throw custom business exceptions.

**Comment by ardalis** [comment]
Can you make your own Guards? The README shows how and they're really easy.

```csharp
// Using the same namespace will make sure your code picks up your 
// extensions no matter where they are in your codebase.
namespace Ardalis.GuardClauses
{
    public static class FooGuard
    {
        public static void Foo(this IGuardClause guardClause, string input, string parameterName)
        {
            if (input?.ToLower() == "foo")
                throw new ArgumentException("Should not have been foo!", parameterName);
        }
    }
}

// Usage
public void SomeMethod(string something)
{
    Guard.Against.Foo(something, nameof(something));
}

```

**Reply by UnleashedGH** [reply]
It would still be nice if the library offered overloaded methods with a CustomException Parameter.
in our application, we _only_ use custom exceptions

thus, it requires us to create custom extensions for all types of functions, and at that point it would be the equivalent of re-creating the library

**Reply by ardalis** [reply]
So you just want one more optional parameter of type Exception?

**Reply by UnleashedGH** [reply]
yes

**Reply by alifaaz** [reply]
@UnleashedGH @ardalis  i have same requirement like @UnleashedGH

**Reply by ardalis** [reply]
Sounds like something we can do for the next major release. Anybody want to do a PR?

**Reply by UnleashedGH** [reply]
for adding the new param, should new extension functions (overloads) be created with the custom exception parameter as required

or, should the existing functions be modified with the new param as optional?

**Reply by ardalis** [reply]
New param optional seems like less code to maintain. I'd go for that unless there's a reason not to (like any method that already accepts a params array at the end, which I haven't checked to see if we have).
